wizard/src/utils/obs_generation.ts

```diff
@@ -0,0 +1,146 @@
+  Observation,
+  RowObservations,
+} from "../types";
+
+// Helper maps used while generating observaitons.
+interface ObsGenMaps {
+  // Column Index -> mapping thing found in header for VALUE.
+  valCol2Hdr: Map<number, MappedThing>;
+  // Mapped thing -> column index (for non-VALUE)
+  thing2Col: Map<MappedThing, number>;
+  // Mapped thing -> constant value (for non-VALUE)
+  thing2Const: Map<MappedThing, string>;
+}
+
+/**
+ * Given a mapping and CSV Data computes the observations for every row
+ * corresponding to |rowsForDisplay|.
+ * ASSUMES: checkMappings() returns success on |mappings|
+ */
+export function generateRowObservations(
+  mappings: Mapping,
+  csvData: CsvData
+): RowObservations {
+  // Compute ObsGenMaps first.
+  const obsGenMaps: ObsGenMaps = {
+    valCol2Hdr: new Map<number, MappedThing>(),
+    thing2Col: new Map<MappedThing, number>(),
+    thing2Const: new Map<MappedThing, string>(),
+  };
+  for (const mthing of Array.from(mappings.keys())) {
+    const mval: MappingVal = mappings.get(mthing);
+    if (mval.type === MappingType.COLUMN_HEADER) {
+      for (const hdr of mval.headers) {
+        obsGenMaps.valCol2Hdr.set(hdr.columnIdx, mthing);
+      }
+    } else if (mval.type == MappingType.COLUMN) {
+      if (mthing == MappedThing.VALUE) {
+        obsGenMaps.valCol2Hdr.set(mval.column.columnIdx, mthing);
+      } else {
+        obsGenMaps.thing2Col.set(mthing, mval.column.columnIdx);
+      }
+    } else if (mval.type == MappingType.CONSTANT) {
+      obsGenMaps.thing2Const.set(mthing, mval.constant);
+    }
+  }
+
+  const rowObs: RowObservations = new Map();
+  for (const idx of Array.from(csvData.rowsForDisplay.keys())) {
+    const row = csvData.rowsForDisplay.get(idx);
+    const obs = generateObservationsInRow(
+      row,
+      csvData.orderedColumns,
+      obsGenMaps
+    );
+    if (obs.length > 0) {
+      rowObs.set(idx, obs);
+    }
+  }
+  return rowObs;
+}
+
+/**
+ * Given an Observation that was returned from generateRowObservations(), returns
+ * a human-readable string.
+ */
+export function observationToString(obs: Observation): string {
+  const sv = obs.get(MappedThing.STAT_VAR);
+  const place = obs.get(MappedThing.PLACE);
+  const date = obs.get(MappedThing.DATE);
+  const val = obs.get(MappedThing.VALUE);
+  let outStr = `Value of ${sv} for ${place} in ${date} is ${val}`;
+  if (obs.has(MappedThing.UNIT)) {
+    outStr += ` ${obs.get(MappedThing.UNIT)}`;
+  }
+  return outStr;
+}
+
+function generateObservationsInRow(
+  row: Array<string>,
+  header: Array<Column>,
+  obsGenMaps: ObsGenMaps
+): Array<Observation> {
+  const obsList = new Array<Observation>();
+  for (const valColIdx of Array.from(obsGenMaps.valCol2Hdr.keys())) {
+    const hdrThing = obsGenMaps.valCol2Hdr.get(valColIdx);
+    const obs: Observation = new Map();
+    // TODO: Consider if we want to do some cleaning and checking for non-numeric value.
+    if (row[valColIdx] != null && row[valColIdx] !== "") {
```

Comment:
can actually just use the lodash function `_.isEmpty()`: https://lodash.com/docs/4.17.15#isEmpty

Reply:
Nice tip!